### ISSUE TITLE
User-creatable tokens for authentication

### ISSUE THREAD

**Issue by AlexGustafsson** [authelia/authelia#2023]
### Use Case

When protecting API-driven services such as a Docker registry, a user may not want to supply their password via a command line application, a third-party UI or the like. Furthermore, a user may want several applications to authenticate on their behalf - with the ability to revoke access without changing their password. In cases like these, the ability to create an API token which allow programs act on the user's behalf is a really helpful feature.

Currently, a workaround is to create a new user such as `my-user-api-token` with a token for a password. This does not scale well and does not let users be in control of their situation.

### Description

By allowing users to create API tokens via the login-portal, they are empowered to access services the way they want to. A new menu in the portal UI may be added to facilitate the creation of tokens. The tokens, which may be assigned a name to help the user describe them, could be stored in the configured session store.

Practically, these tokens would be used instead of the password when accessing resources. My use cases only require them to work for Basic Auth / Authorization / Proxy Authorization endpoints, which I assume is natural for such a feature. Furthermore, in order to allow admins to control what can and cannot be done with the tokens, one may add a third policy type alongside `one_factor` and `two_factor` - `token`. A token should not be able to be used to access the portal, as that would make a holder of a token to create more tokens.

Although the feature(s) proposed is approaching that of an API gateway, I believe it could be in scope of this project as well.

**Comment by AlexGustafsson** [comment]
Given that this is a direction that this project wants to take, I'd be able to take a look at implementing it in the time to come.

I have some areas of concerns regarding the feature:

* Is there a reason to store the tokens in the backing store / configuration database instead of in the session store?
* What should a token be allowed to do and what should it not? How is that controlled by an admin?
* Should an admin be able to toggle this on/off globally or on a per-user / per-group basis?

**Comment by james-d-elliott** [comment]
Hey @AlexGustafsson thanks for your suggestion! This is something we had in mind to do. Our current priorities are finalizing the OIDC implementation, and user/admin control panels. However I believe those control panels will enable us to better support something like this. Keeping in mind this is just my personal opinions as one of the maintainers, @clems4ever  and @nightah may have other opinions or ideas, or you may have valid counter points which will also be considered. The important thing is we come up with a clear design before you put effort into it.

>     * What should a token be allowed to do and what should it not? How is that controlled by an admin?

The *primary* issue of concern that needs to be ensured in my mind is that the tokens cannot give users more access than they already have. We would probably have to implement a method in the authorization package specifically to handle these tokens, to ensure this occurs on every request, as the token may have a longer lifetime than the users access to a resource (i.e. change in groups). We could realistically add something to ACL rules to apply a specific policy for tokens. Two examples:

```yaml
access_control:
  default_policy: deny
  rules:
  - domain: example.com
    policy: one_factor
    allow_token: true
```

```yaml
access_control:
  default_policy: deny,token_deny
  rules:
  - domain: example.com
    policy: one_factor,token_allow
```

>     * Is there a reason to store the tokens in the backing store / configuration database instead of in the session store?

We currently must store them in the SQL storage, the redis storage is 100% controlled by the fasthttp/session package which only allows storing things against individual sessions. This would make it impossible to lookup the tokens in situations where they are most relevant since the cookie linking the client (browser) to the session in redis would not be available, thus the scope would be different.

>     * Should the contents of the tokens be decided by the user or generated by Authelia?

I do not believe it is good security practice to allow tokens to be provided by end users, they should be provided by Authelia either as random alphanumeric strings or as a signed JWT.

>     * Should an admin be able to toggle the ability to create tokens globally or on a per-user / per-group basis?
>     * Should tokens be set to expire globally? Should the user be in charge of this?

I think allowing admins to set a maximum lifetime for a token (0 for disable, -1 or lower for unlimited) *and* giving users an option provided it doesn't exceed this maximum is the best solution of both worlds. So for example if admins set it to 30d then users can pick anything from 1s to 30d, but no more than 30d. This would also take care of admins being able to toggle global token creation.

For individual user/group creation this is something we can easily add later as an enhancement. I think we need to probably have an internal discussion for long term intent regarding groups, I think it makes sense for us to consider a differentiation between internal groups and authentication groups. For example it would be beneficial to have groups that are intended for mapping internal rules, and then *map external groups* to those groups. This would allow multi-auth backends. Though maybe that can just be redone when/if that occurs.

**Comment by AlexGustafsson** [comment]
So to summarize a bit from above and to make sure I've understood you correctly, the MVP is basically the following:

* Add support for specifying more than one default policy in the configuration (of course making sure that they don't contradict each other such as allow, deny). Like with other configuration variables, use `weak` to be backwards compatible with current configurations, but allow two to be specified.
* Add support for specifying more than one policy in a rule (same as above).
* Add two new policies, `allow_token` and `deny_token` that may be used in the configuration file.
* Add configuration for maximum token lifetime.
* Create an API to allow authenticated users to create tokens to themselves. It would take two parameters, name and lifetime (which may not exceed the maximum token lifetime). The server creates and signs a JWT, id and stores them together with the revocation timestamp based on the lifetime and name in the database. The token is returned in the response.
* Create an API to allow authenticated users to list tokens owned by them. The API returns the id, name and revocation time for each token.
* Create an API to allow authenticated users to delete tokens owned by them. It would take the token id as a parameter.
* Add support for the `Authorization` header in the `verify` handler. If the header is set, expect the `Bearer <token>` pattern and get the token. Verify the token. Look up and return the token's owner. This would work much like the existing basic auth handler. Use a new `Token` authentication level.

The APIs would require the highest possible authentication level for the user and explicitly disallow the new token authentication level.

The API parts are probably better explained via an OpenAPI schema, but I'd like to arrive at a common understanding and goal first.

In terms of UI, if the control panel is reasonably close on the roadmap, it would perhaps make sense to not supply a UI for the functionality just yet, but just getting the API done as a first stretch. If it's far off, it might make sense to have a temporary UI to provide the user with the functionality.

Also, I'm a bit undecided if a JWT is warranted or not. As Authelia wouldn't use them in a stateless way anyway (in order to support revocation by the user), one may as well go with a random token. The only positive I find with JWTs in this context would be a way to easily expose to the user when the token expires (`exp` claim). Perhaps the `jti` (JWT ID) claim could be added as well to supply the user with the token's id. 

Lastly, does it ever make sense to support the tokens in turn for the user's password / first factor? Some parts explained here could easily be extended to support the functionality proposed in #1387. Or should that be done as a new policy as well, such as `shared_token`? If the functionality should be shared, one may allow users to create tokens for certain resources, make them usable only once etc. It's perhaps not a prioritization for this feature, but it may be worth considering.

**Comment by nsauter** [comment]
Is there anything new here? Are you planning to implement that or is this still on hold?

**Comment by james-d-elliott** [comment]
It's likely landing in 4.38.0 via a user downloadable CLI. 

See https://github.com/authelia/authelia/pull/5916

**Comment by ThinkChaos** [comment]
Thanks for working on this!  
I read through the [new docs](https://658e93ad2887e400089d694d--authelia-staging.netlify.app/integration/openid-connect/oauth-2.0-bearer-token-usage/) but didn't see any mention of the CLI. Are there any public details about that?  
Basically I'm wondering what the flow for a user to create a token would look like. Is it integrated in the main authelia CLI? Can an admin easily generate a token for a user? Similar to how one can generate a password hash maybe?

**Comment by james-d-elliott** [comment]
There are no public instructions yet. The CLI will be completely separate,, will only have this feature or intent to begin with, will need a fairly basic config, it will be installable via the go toolchain making it accessible  on every OS that go is installable on (i.e. a lot). It would feel fairly similar to any other CLI that does this, you're given a link, perform the flow, and you're returned a token and a YAML config which has the token and refresh token if applicable. Admins will NOT be able to generate tokens for users unless they have their credentials.
